crates/bevy_ecs/src/schedule_v3/graph/utils.rs

```diff
@@ -0,0 +1,197 @@
+    topological_order: &[V],
+) -> CheckGraphResults<V>
+where
+    V: NodeTrait + Debug,
+{
+    if graph.node_count() == 0 {
+        return CheckGraphResults::default();
+    }
+
+    let n = graph.node_count();
+    let mut map = HashMap::with_capacity(n);
+    let mut tsorted = DiGraphMap::<V, ()>::new();
+    // iterate nodes in topological order
+    for (i, &node) in topological_order.iter().enumerate() {
+        map.insert(node, i);
+        tsorted.add_node(node.clone());
+        // insert nodes as successors to their predecessors
+        for pred in graph.neighbors_directed(node, Direction::Incoming) {
+            tsorted.add_edge(pred, node, ());
+        }
+    }
+
+    let mut tred = DiGraphMap::<V, ()>::new();
+    let mut tcls = DiGraphMap::<V, ()>::new();
+
+    let mut connected = HashSet::new();
+    let mut disconnected = HashSet::new();
+    let mut transitive_edges = Vec::new();
+
+    let mut visited = FixedBitSet::with_capacity(n);
+    let mut reachable = FixedBitSet::with_capacity(n * n);
+
+    // iterate nodes in topological order
+    for node in tsorted.nodes() {
+        tred.add_node(node);
+        tcls.add_node(node);
+    }
+
+    // iterate nodes in reverse topological order
+    for a in tsorted.nodes().rev() {
+        let index_a = *map.get(&a).unwrap();
+        // iterate their successors in topological order
+        for b in tsorted.neighbors_directed(a, Direction::Outgoing) {
+            let index_b = *map.get(&b).unwrap();
+            debug_assert!(index_a < index_b);
+            if !visited[index_b] {
+                // edge <a, b> is not redundant
+                tred.add_edge(a, b, ());
+                tcls.add_edge(a, b, ());
+                reachable.set(index(index_a, index_b, n), true);
+
+                let successors = tcls
+                    .neighbors_directed(b, Direction::Outgoing)
+                    .collect::<Vec<_>>();
+                for c in successors.into_iter() {
+                    let index_c = *map.get(&c).unwrap();
+                    debug_assert!(index_b < index_c);
+                    if !visited[index_c] {
+                        visited.set(index_c, true);
+                        tcls.add_edge(a, c, ());
+                        reachable.set(index(index_a, index_c, n), true);
+                    }
+                }
+            } else {
+                transitive_edges.push((a, b));
+            }
+        }
+
+        visited.clear();
+    }
+
+    for i in 0..(n - 1) {
+        // reachable is upper triangular because the nodes are in topological order
+        for index in index(i, i + 1, n)..=index(i, n - 1, n) {
+            let (a, b) = row_col(index, n);
+            let pair = (topological_order[a], topological_order[b]);
+            if !reachable[index] {
+                disconnected.insert(pair);
+            } else {
+                connected.insert(pair);
+            }
+        }
+    }
+
+    // Fill diagonal.
```

Comment:
Why?

Reply:
Some graph algs expect that nodes count as being connected to themselves, but I've commented this part out for now since it wasn't necessary for the system set condition handling.